### PR TITLE
Smart linking in web apps

### PR DESCRIPTION
**Pull request by orangejenny**
Re-opening https://github.com/dimagi/commcare-hq/pull/30441 which was accidentally merged.

## Product Description
https://dimagi-dev.atlassian.net/browse/USH-1292

## Technical Summary
See https://github.com/dimagi/formplayer/pull/1010

## Feature Flag
Data registries

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

Little to none - this is mostly web apps UI where test coverage isn't great.

### QA Plan

Requesting QA

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

**Comment by orangejenny** [comment]
Pinging for review - @snopoke you already reviewed most of this in https://github.com/dimagi/commcare-hq/pull/30441 and various sub-PRs.

**Comment by snopoke** [comment]
> Pinging for review - @snopoke you already reviewed most of this in #30441 and various sub-PRs.

I'll do another pass tomorrow

**Comment by snopoke** [comment]
@orangejenny there are some branch conflicts

**Comment by orangejenny** [comment]
@snopoke Thanks, fixed.

File: corehq/apps/app_manager/suite_xml/post_process/endpoints.py [hunk 1]
```diff
@@ -61,13 +61,14 @@ def _make_session_endpoint(self, module, form=None):
...
-            id=endpoint_id,
-            arguments=[Argument(id=i) for i in argument_ids],
-            stack=stack,
-        )
+        kwargs = {
+            "id": endpoint_id,
+            "arguments": [Argument(id=i) for i in argument_ids],
+            "stack": stack,
+        }
+        return SessionEndpoint(**kwargs)
```
Comment (snopoke) [inline corehq/apps/app_manager/suite_xml/post_process/endpoints.py:69]: nit: this change seems unnecessary

File: corehq/apps/app_manager/suite_xml/post_process/remote_requests.py [hunk 2]
```diff
@@ -254,15 +265,74 @@ def build_query_prompts(self):
...
+                ),
+            ))
+            stack.add_frame(frame)
+        frame = PushFrame(if_clause=rewind_if)
         frame.add_rewind(QuerySessionXPath(self.case_session_var).instance())
         stack.add_frame(frame)
         return stack
 
+    def get_smart_link_function(self):
+        app_id = self.app.upstream_app_id if is_linked_app(self.app) else self.app.origin_id
```
Comment (snopoke) [inline corehq/apps/app_manager/suite_xml/post_process/remote_requests.py:290]: does smart links to other domains work at all for non-linked apps?

File: corehq/apps/app_manager/suite_xml/post_process/remote_requests.py [hunk 3]
```diff
@@ -254,15 +265,74 @@ def build_query_prompts(self):
...
+        url = absolute_reverse("session_endpoint", args=["---", app_id, self.module.session_endpoint_id])
+        prefix, suffix = url.split("---")
+        params = ""
+        argument_ids = self.endpoint_argument_ids
+        if argument_ids:
+            params = f", '?{argument_ids[-1]}=', ${argument_ids[-1]}"
+            for argument_id in argument_ids[:-1]:
+                params += f", '&{argument_id}=', ${argument_id}"
```
Comment (snopoke) [inline corehq/apps/app_manager/suite_xml/post_process/remote_requests.py:298]: this is hard to grok but a comment with an example above would help:
```
# URL params need to be evaluated in xpath:
#    concat('?arg1=', $arg1, '&arg2=', $arg2)
#  after xpath eval:
#    '?arg1=va1&arg2=val2'
```

Side note: an xpath function to build URLs that also encoded params would be great

File: corehq/apps/app_manager/suite_xml/post_process/remote_requests.py [hunk 4]
```diff
@@ -254,15 +265,74 @@ def build_query_prompts(self):
...
     def build_stack(self):
         stack = Stack()
-        frame = PushFrame()
+        rewind_if = None
+        if module_uses_smart_links(self.module):
+            user_domain_xpath = session_var(COMMCARE_PROJECT, path="user/data")
+            # For case in same domain, do a regular case claim rewind
+            rewind_if = self._get_case_domain_xpath().eq(user_domain_xpath)
+            # For case in another domain, jump to that other domain
+            frame = PushFrame(if_clause=self._get_case_domain_xpath().neq(user_domain_xpath))
```
Comment (snopoke) [inline corehq/apps/app_manager/suite_xml/post_process/remote_requests.py:274]: nit: perhaps this is personal preference and I know that these are logically the same but I think this is clearer that one is the inverse of the other:
```suggestion
            frame = PushFrame(if_clause=XPath.not_(rewind_if))
```

File: corehq/apps/app_manager/tests/test_suite_remote_request.py [hunk 5]
```diff
@@ -31,6 +38,51 @@
...
+
+    def testSmartLinkVariables(self):
+        vars = self.request_factory.get_smart_link_variables()
+        self.assertEqual([v.name for v in vars], ['domain', 'case_id', 'case_id_leaf'])
+        self.assertEqual([v.xpath.function for v in vars], [
+            "instance('results')/results/case[@case_id=instance('commcaresession')"
+            "/session/data/search_case_id]/commcare_project",
+            "instance('commcaresession')/session/data/case_id",
+            "instance('commcaresession')/session/data/search_case_id",
+        ])
```
Comment (snopoke) [inline corehq/apps/app_manager/tests/test_suite_remote_request.py:83]: this was quite confusing until I saw the missing comma, even then it's hard to read
```suggestion
        session_case_id = "instance('commcaresession')/session/data/search_case_id"
        self.assertEqual([v.xpath.function for v in vars], [
            f"instance('results')/results/case[@case_id={session_case_id}]/commcare_project",
            "instance('commcaresession')/session/data/case_id",
            "instance('commcaresession')/session/data/search_case_id",
        ])
```

File: corehq/apps/app_manager/suite_xml/xml_models.py [hunk 6]
```diff
@@ -463,13 +463,20 @@ def add_frame(self, frame):
...
+
+
 class Argument(IdNode):
     ROOT_NAME = 'argument'
 
 
 class SessionEndpoint(IdNode):
     ROOT_NAME = 'endpoint'
 
+    command_id = StringField('@command_id')
```
Comment (snopoke) [inline corehq/apps/app_manager/suite_xml/xml_models.py:479]: it this still used? I don't see it in any of the test XML and thought I recalled you removing it.

**Review by snopoke — APPROVED** [review]
no blockers, just small comments

File: corehq/apps/app_manager/suite_xml/post_process/remote_requests.py (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/app_manager/suite_xml/post_process/remote_requests.py:290]: Only for non-linked apps that are upstream apps - you can link between a corresponding upstream & downstream app.

File: corehq/apps/app_manager/suite_xml/post_process/remote_requests.py (hunk 4 above)
Comment (orangejenny) [reply corehq/apps/app_manager/suite_xml/post_process/remote_requests.py:274]: thanks, that's much more readable

File: corehq/apps/app_manager/suite_xml/post_process/remote_requests.py (hunk 3 above)
Comment (orangejenny) [reply corehq/apps/app_manager/suite_xml/post_process/remote_requests.py:304]: 260ef14a3c0

Like a function that would build the URL on the CC side? That'd be cool.

File: corehq/apps/app_manager/suite_xml/post_process/endpoints.py (hunk 1 above)
Comment (orangejenny) [reply corehq/apps/app_manager/suite_xml/post_process/endpoints.py]: bfc6710f02615389d2e27a7e1bfc8438d66c778e

File: corehq/apps/app_manager/suite_xml/xml_models.py (hunk 6 above)
Comment (orangejenny) [reply corehq/apps/app_manager/suite_xml/xml_models.py]: oops, bfc6710f02615389d2e27a7e1bfc8438d66c778e

File: corehq/apps/app_manager/suite_xml/post_process/remote_requests.py [hunk 7]
```diff
@@ -254,15 +265,74 @@ def build_query_prompts(self):
...
     def build_stack(self):
         stack = Stack()
-        frame = PushFrame()
+        rewind_if = None
+        if module_uses_smart_links(self.module):
+            user_domain_xpath = session_var(COMMCARE_PROJECT, path="user/data")
+            # For case in same domain, do a regular case claim rewind
+            rewind_if = self._get_case_domain_xpath().eq(user_domain_xpath)
+            # For case in another domain, jump to that other domain
+            frame = PushFrame(if_clause=XPath.not_(rewind_if))
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/suite_xml/post_process/remote_requests.py:275]: fixed in a different commit

**Pull request merged**